mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLExecutionDAO.java

```diff
@@ -0,0 +1,761 @@
+
+		if (current >= limit) {
+			Monitors.recordTaskRateLimited(task.getTaskDefName(), limit);
+			return true;
+		}
+
+		logger.info("Task execution count for {}: limit={}, current={}", task.getTaskDefName(), limit, getInProgressTaskCount(task.getTaskDefName()));
+
+		String taskId = task.getTaskId();
+
+		List<String> tasksInProgressInOrderOfArrival = findAllTasksInProgressInOrderOfArrival(task, limit);
+
+		boolean rateLimited = !tasksInProgressInOrderOfArrival.contains(taskId);
+
+		if (rateLimited) {
+			logger.info("Task execution count limited. {}, limit {}, current {}", task.getTaskDefName(), limit, getInProgressTaskCount(task.getTaskDefName()));
+			Monitors.recordTaskRateLimited(task.getTaskDefName(), limit);
+		}
+
+		return rateLimited;
+	}
+
+	@Override
+	public void updateTasks(List<Task> tasks) {
+		withTransaction(connection -> tasks.forEach(task -> updateTask(connection, task)));
+	}
+
+	@Override
+	public void addTaskExecLog(List<TaskExecLog> log) {
+		indexer.add(log);
+	}
+
+	@Override
+	public void removeTask(String taskId) {
+		Task task = getTask(taskId);
+
+		if(task == null) {
+			logger.warn("No such Task by id {}", taskId);
+			return;
+		}
+
+		String taskKey = task.getReferenceTaskName() + "" + task.getRetryCount();
+
+		withTransaction(connection -> {
+			removeScheduledTask(connection, task, taskKey);
+			removeWorkflowToTaskMapping(connection, task);
+			removeTaskInProgress(connection, task);
+			removeTaskData(connection, task);
+		});
+	}
+
+	@Override
+	public Task getTask(String taskId) {
+		String taskJsonStr = getWithTransaction(c -> c.createQuery(GET_TASK).addParameter("taskId", taskId).executeScalar(String.class));
+		return taskJsonStr != null ? readValue(taskJsonStr, Task.class) : null;
+	}
+
+	@Override
+	public List<Task> getTasks(List<String> taskIds) {
+		if (taskIds.isEmpty()) return Lists.newArrayList();
+
+		List<String> taskIdParameters = taskIds.stream().map(taskId -> format("'%s'", taskId)).collect(Collectors.toList());
+
+		return getWithTransaction(c -> c.createQuery(format(GET_TASKS, join(taskIdParameters, ","))).executeScalarList(String.class))
+				.stream()
+				.filter(Objects::nonNull)
+				.map(taskJsonStr -> readValue(taskJsonStr, Task.class))
+				.collect(Collectors.toList());
+	}
+
+	@Override
+	public List<Task> getPendingTasksForTaskType(String taskName) {
+		Preconditions.checkNotNull(taskName, "task name cannot be null");
+
+		List<String> taskIds = getWithTransaction(connection -> connection.createQuery(GET_ALL_IN_PROGRESS_FOR_TASK_TYPE)
+				.addParameter("taskDefName", taskName)
+				.executeScalarList(String.class));
+
+		return getTasks(taskIds);
+	}
+
+	@Override
+	public List<Task> getTasksForWorkflow(String workflowId) {
+		List<String> taskIds = getWithTransaction(c -> c.createQuery(GET_TASKS_FOR_WORKFLOW)
+				.addParameter("workflowId", workflowId)
+				.executeScalarList(String.class));
+
+		return getTasks(taskIds);
```

Comment:
Why does this have to be done in two transactions?

Reply:
not a very good reason, mostly because `getTasks` is a public DAO method and I cant change the signature

I will create an internal getTasks that takes a connection and reuse it in both places